flang/lib/Lower/IO.cpp

```diff
@@ -196,6 +196,144 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
   builder.setInsertionPointToStart(&ifOp.thenRegion().front());
 }
 
+/// Retrieve or generate (the constant parts of) information for NAMELIST
+/// group `symbol`.  Group object descriptor addresses are inserted at
+/// runtime.  The data structure containing the information is defined in
+/// runtime header file namelist.h as:
+///
+///     class NamelistGroup {
+///     public:
+///       struct Item {
+///         const char *name; // NUL-terminated lower-case
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
+///
+/// Return the Item array and the NamelistGroup data structure.
+static std::pair<fir::GlobalOp, fir::GlobalOp>
+getNamelistGroup(Fortran::lower::AbstractConverter &converter,
+                 const Fortran::semantics::Symbol &symbol) {
+  auto &builder = converter.getFirOpBuilder();
+  auto mangledGroupName = converter.mangleName(symbol);
+  auto listName = mangledGroupName + ".list";
+  if (auto group = builder.getNamedGlobal(mangledGroupName)) {
+    auto list = builder.getNamedGlobal(listName);
+    assert(list && "missing namelist group list");
+    return {list, group};
+  }
+  const auto &details = symbol.get<Fortran::semantics::NamelistDetails>();
+  auto loc = converter.getCurrentLocation();
+  auto *context = builder.getContext();
+  auto linkOnce = builder.createLinkOnceLinkage();
+  auto idxTy = builder.getIndexType();
+  auto sizeTy = builder.getIntegerType(8 * sizeof(std::size_t));
+  auto charPtrTy = fir::PointerType::get(builder.getIntegerType(8));
+  auto descPtrTy =
+      fir::PointerType::get(fir::BoxType::get(mlir::NoneType::get(context)));
+  auto listTy = fir::SequenceType::get(
+      details.objects().size(),
+      mlir::TupleType::get(context, {charPtrTy, descPtrTy}));
+  auto groupTy = mlir::TupleType::get(
+      context, {charPtrTy, sizeTy, fir::PointerType::get(listTy)});
+  auto stringAddress = [&](const Fortran::semantics::Symbol &symbol,
+                           bool wantResult = false) {
+    auto global =
+        createStringLiteral(builder, loc, symbol.name().ToString() + '\0');
+    return wantResult
+               ? builder.createConvert(loc, charPtrTy, fir::getBase(global))
+               : mlir::Value{};
+  };
+  stringAddress(symbol);
+  for (const Fortran::semantics::Symbol &s : details.objects())
+    stringAddress(s);
```

Comment:
If the goal of `stringAddress` gymnastic  is to order the global creations according to their dependencies, wouldn't it be simpler to put the character globals in a smallVector an later looping over it  ?

```
llvm::SmallVector<mlir::Value> nameAddresses;
for (const Fortran::semantics::Symbol &s : details.objects()) {
  auto global =  createStringLiteral(builder, loc, s.name().ToString() + '\0');
  nameAddresses.push_back(builder.createConvert(loc, charPtrTy, fir::getBase(global)));
}
auto listFunc = [&](Fortran::lower::FirOpBuilder &builder) {
  // ...
 for (auto name : nameAddresses)
   list = builder.create<fir::InsertValueOp>(loc, listTy, list, name, idx);
 // ...
}
```

Reply:
`stringAddress` has been simplified.